fixed is_btrfs check, added warning

h3po opened the pull request.

anthonyrisinger commented on btrfs_hook:
```diff
@@ -68,8 +68,8 @@ BTRFS_DIR_ROLLBACK="${BTRFS_DIR_ROLLBACK%/}"
-    [ "$(blkid -s TYPE -o value "$root" )" != "btrfs" ] &&
-        [ "$(blkid -s TYPE -o value -lt "$root" )" != "btrfs" ]
+    [ "$(blkid -s TYPE -o value \"$root\" )" != "btrfs" ] &&
+        [ "$(blkid -s TYPE -o value -lt \"$root\" )" != "btrfs" ]
```
How was this failing? What was the error message? This seems wrong to me -- quoting the `"` will make it part of the value.

h3po commented:
the check was just failing, i just quickly added a full blkid output before the check and determined that it must be the unbalanced quotes. with the patch, the first check correctly returns 1. afaik, not quoting them breaks the eval. i don't see why there should be quotes anyway.

h3po commented:
you are correct, this works only with the unescaped quotes. in my init however, it works only with the escapes. i have no idea what's going on. there's no quotes in my $root.
    #!/bin/bash
    root=/dev/mapper/root_ssd_crypt
    echo [ "$(blkid -s TYPE -o value $root )" ]
    echo [ "$(blkid -s TYPE -o value "$root" )" ]
    echo [ "$(blkid -s TYPE -o value \"$root\" )" ]

anthonyrisinger commented:
That is really weird. You could try adding `break=y` to kernel cmdline to explore... shell should be `ash` IIRC. Quotes are generally only needed to control word-splitting, and it's possible (though certainly uncommon) to have whitespace in the root device path. This can be controlled with something like `IFS=$'\n'`, which I normally do in my own scripts (I personally think it's stupid things are split on spaces), but I just didn't want to deal with it here. Checking for btrfs device is outdated anyway, should probably just go ahead and try it. See #16.

anthonyrisinger commented:
I'd be curious in knowing the value of `$root` in your case.

h3po commented:
At the point that the hook runs, my $root is /dev/mapper/root_ssd_crypt. I did actually put btrfs_info in all possible places of the hook, that's how i discovered there was no output (except config file loaded) because the check failed.
grub-mkconfig populates cmdline with root=UUID=... and rootflags=subvol=@root and that gets rewritten by one of the earlier hooks (I guess "filesystems").
I might poke around a little more, maybe embed that testscript I posted into init.

h3po commented:
It has to be something with the busybox ash. I've copied the root_is_btrfs into a test script and played around with it (break=postmount) in the init shell, the check only works with the escaped quotes. I have no background in shell scripting, so I don't know how to debug this further.

anthonyrisinger commented:
Try moving the btrfs hook after the crypt hooks. I think this is related to #16, the order should not matter. What does your `mkinitcpio.conf` look like (hooks line)?

h3po commented:
If the btrfs hook had been before the crypt hooks, of course it couln't have detected the filesystem type on the (non-existent) mapper. To be clear, my boot process was never broken. btrfs_advanced just called the default mount handler, which then mapped my default subvolume @root. I was just not getting the option to rollback etc.

When I was fiddling with btrfs_advanced initially, my hook line was

```
base udev autodetect modconf block filesystems keyboard fsck encrypt encrypt2 btrfs_advanced
```

With a duplicate encrypt hook as per https://wiki.archlinux.org/index.php/Dm-crypt/Specialties#Modifying_the_encrypt_hook_for_multiple_partitions
Now that btrfs_advanced works for me (with the patch I submitted), I have added netconf+dropbear and replaced encrypt+encrypt2 with mkinitcpio-multiencrypt. This I am currently editing to output the necessary information to files that can then be used by a shell script similar to cryptsetup_shell from mkinitcpio-utils to unlock my two disks remotely. At this stage, my settings are:

```
/etc/mkinitcpio.conf:
MODULES="loop"
BINARIES="nano"
FILES="/usr/share/terminfo/l/linux"
HOOKS="base udev autodetect modconf block filesystems keyboard keymap netconf dropbear multiencrypt_ssh btrfs_advanced"

/etc/default/grub:
GRUB_CMDLINE_LINUX_DEFAULT="break=postmount nomodeset ip=192.168.0.2::192.168.0.1:255.255.255.0::eth1:none cryptdevices=\"UUID=8a2d1339-abcd-460f-abcd-48a469744a7a:root_ssd_crypt:allow-discards;UUID=e71c2533-abcd-4089-abcd-c80aa325812c:root_hdd_crypt\" break=postmount"
-> resulting /boot/grub/grub.cfg cmdline:
linux   /vmlinuz-linux root=UUID=d941915e-d126-41b2-8877-c34f615cd8fb rw rootflags=subvol=@root  break=postmount nomodeset ip=192.168.0.2::192.168.0.1:255.255.255.0::eth1:none cryptdevices="UUID=8a2d1339-abcd-460f-abcd-48a469744a7a:root_ssd_crypt:allow-discards;UUID=e71c2533-abcd-4089-abcd-c80aa325812c:root_hdd_crypt" break=postmount

/etc/default/btrfs_advanced:
BTRFS_DIR_ACTIVE="/@root"
BTRFS_DIR_SNAPSHOT="/@root_snapshots"
BTRFS_DIR_ROLLBACK="/__rollback"
BTRFS_ROLLBACK_KERNEL=false
BTRFS_ENABLE_DEGRADED=true
```

I have tested the rollback functionality successfully. To test degraded mode, I'll first have to duplicate my /boot to the second disk.
Maybe at some point I'll figure out how to log the stdout, so I could give you the init console output.

h3po commented:
here's a vmware vm with two luks drives and multiencrypt+btrfs_advanced hooks. pw root:root, user:user and the luks password is "crypt". it has the same exact problem, the btrfs check only works when i escape the quotes.
http://212.227.251.207:8080/arch-test.zip
